Experiment Searching: Search for experiments by name in the UI

ggliem opened the pull request:
## What changes are proposed in this pull request?

Experiment Searching: Added option to search for experiments by name to the UI. This is especially useful when e.g. collaborating within a team and creating a large number of experiments. A case-insensitive 'contains' search is performed when searching for an experiment.
This feature has also been requested in https://github.com/mlflow/mlflow/issues/1431.
Please refer to the following gif which demonstrates the change:

![search_experiment](https://user-images.githubusercontent.com/15968047/72704764-87136e80-3b0e-11ea-950b-eec44cb0c05d.gif)



## How is this patch tested?

- Added unit tests to ExperimentListView.test.js
- Manual testing


## Release Notes

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

Added option to search for experiments by name to the UI.

### What component(s) does this PR affect?

- [X] UI
- [ ] CLI
- [ ] API
- [ ] REST-API
- [ ] Examples
- [ ] Docs
- [ ] Tracking
- [ ] Projects
- [ ] Artifacts
- [ ] Models
- [ ] Scoring
- [ ] Serving
- [ ] R
- [ ] Java
- [ ] Python

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

ggliem committed:
Add search input to experiment list

Added search input field to experiment list view.

ggliem committed:
Filter experiment list based on input from search field

Added filtering functionality.
When using the experiment search field, only experiments that contain
the searched value will be shown.

ggliem committed:
Add unit tests

Added unit tests, testing the added search functionality.

ggliem committed:
Format Fix

Zangr commented:
cc @gioa

ggliem committed:
Align with naming conventions

gioa reviewed (commented):
@ggliem Thank you for working on it! 
The gif looks good. One suggestion is to change the focus style of the search box from red to blue (e.g., #258BD2) since red often indicates an error.  Thank you!

cc @Zangr for the code review.

ggliem committed:
Update focus style of input field

Changed focus style of search box from red to blue since red often indicates an error.

ggliem commented:
> @ggliem Thank you for working on it!
> The gif looks good. One suggestion is to change the focus style of the search box from red to blue (e.g., #258BD2) since red often indicates an error. Thank you!

I changed the focus style. This is how it looks like now.
![mlflow_experiment_search](https://user-images.githubusercontent.com/15968047/73721525-4e13f600-46d9-11ea-9e2e-2af294c4d3cb.gif)

One remark: I noticed that the focus style of some of the other input fields (e.g. "Search Runs") also includes the red outline. I don't know whether that's intended in this case. In my opinion, it might make sense to keep these styles consistent. What do you think?

gioa commented:
@ggliem I totally agree! Thank you for fixing the focus style on this one, LGTM!

ggliem commented:
@Zangr Following up on this PR - Have you had time to look into this?

Zangr commented on mlflow/server/js/src/components/ExperimentListView.css:
```diff
@@ -23,10 +23,24 @@
   font-weight: normal;
   display: inline-block;
-  padding-bottom: 6px;
+  /* padding-bottom: 6px; */
```
We can remove it if it's not needed.

Zangr commented on mlflow/server/js/src/components/ExperimentListView.css:
```diff
@@ -47,4 +61,4 @@
   margin-left: 96px;
   cursor: pointer;
-}
+}
```
nit: new line eof

Zangr commented on mlflow/server/js/src/components/ExperimentListView.js:
```diff
@@ -30,11 +31,17 @@ export class ExperimentListView extends Component {
     // 100 for the experiments header and some for bottom padding.
     const experimentListHeight = height - 60 - 100;
+    // get searchInput from state
+    const {searchInput} = this.state;
```
```suggestion
    const { searchInput } = this.state;
```

Zangr commented on mlflow/server/js/src/components/ExperimentListView.js:
```diff
@@ -44,13 +51,25 @@ export class ExperimentListView extends Component {
+            {
+            // filter experiments based on searchInput
+            this.props.experiments.filter((e) => {
+              return e.getName().toLowerCase().includes(searchInput.toLowerCase());
```
nit: unnecessary return statement.
`(e) => e.getName().toLowerCase().includes(searchInput.toLowerCase())`

Zangr commented on mlflow/server/js/src/components/ExperimentListView.js:
```diff
@@ -44,13 +51,25 @@ export class ExperimentListView extends Component {
+                active = parsedExperimentId === this.props.activeExperimentId;
               } else {
-                active = idx === 0;
+                active = parsedExperimentId === 0;
```
This change the default active experiment from the 1st experiment to the experiment with Id=0. Is this what we wanted?

ggliem replied:
Good point. One issue that I encountered was that when selecting the experiment with ID 0 & entering a search term, the first item in the filtered list would be shown as active. Since `0` is a falsy value, the active mark was set where `idx === 0`.
I updated the condition in the if clause to `this.props.activeExperimentId !== undefined` and undid the change that you mentioned.

ggliem committed:
Update default experiment selection & minor refactoring

Zangr committed:
clean, format

Zangr commented:
LGTM
